lib/Parse/ParseDecl.cpp

```diff
@@ -3347,7 +3371,35 @@ bool Parser::isStartOfDecl() {
     if (Tok.isAny(tok::r_brace, tok::eof, tok::pound_endif))
       return true;
 
-    return isStartOfDecl();
+    return isStartOfSwiftDecl();
+  }
+
+  // If we have a decl modifying keyword, check if the next token is a valid
+  // decl start. This is necessary to correctly handle Swift keywords that are
+  // shared by SIL, e.g 'private'.
```

Comment:
Can you give an example?

Reply:
@slavapestov This handles cases like `sil private @foo : @convention(thin) () -> ()`. If we unconditionally treat `private` as the start of a decl, we'll try to parse this as a Swift decl. Should I add this example to the comment?

Reply:
I've gone ahead and elaborated a little on the comment